index.bs

```diff
@@ -773,17 +753,24 @@ The following steps are taken:
 1. Let _stageValue_ be ToString(_stage_).
 1. RejectIfAbrupt(_stageValue_).
 1. If IsValidStageValue(_stageValue_) is *false*, return a promise rejected with a new *RangeError* exception.
+1. Let _stageEntry_ be GetStage(_entry_, _stageValue_).
+1. If _stageEntry_ is *undefined*, return a promise rejected with a new *TypeError* exception.
+1. Perform UpgradeToStage(_entry_, _stageValue_).
 1. Let _p0_ be the result of transforming _result_ with a new pass-through promise.
 1. Let _p1_ be the result of transforming _p0_ with a fulfillment handler that, when called with argument _value_, runs the following steps:
-  1. Let _stageEntry_ be GetStage(_entry_, _stageValue_).
-  1. If _stageEntry_ is *undefined*, throw a new *TypeError*.
-  1. If _stageEntry_.[[Result]] is *undefined*, then
-    1. Set _stageEntry_.[[Result]] to a promise resolved with _value_.
+  1. If _stageValue_ is "instantiate", then:
+    1. Perform ? ExtractDependencies(_entry_, _value_, *undefined*).
+    1. Return the result of transforming RequestSatisfy(_entry_, *undefined*) with a fulfillment handler that, when called, runs the following steps:
+      1. Let _stageEntry_ be GetStage(_entry_, _stageValue_).
+      1. Assert: _stageEntry_ is not *undefined*.
+      1. Fulfill _stageEntry_.[[Result]] with _value_.
   1. Else,
+    1. Let _stageEntry_ be GetStage(_entry_, _stageValue_).
+    1. If _stageEntry_ is *undefined*, throw a new *TypeError*.
     1. Fulfill _stageEntry_.[[Result]] with _value_.
-  1. Perform UpgradeToStage(_entry_, _stageValue_).
 1. Let _pCatch_ be the result of transforming _p1_ with a rejection handler that, when called, runs the following steps:
   1. Set _entry_.[[Error]] to *true*.
+1. If _stageEntry_.[[Result]] is *undefined*, set _stageEntry_.[[Result]] to _p1_.
```

Comment:
trying to eliminate the racing nature of this by taking over the `[[Result]]` when possible.